designs/0005-static-matrices.md

```diff
@@ -0,0 +1,95 @@
+- Feature Name: static_matrices
+- Start Date: April 30th, 2020
+- RFC PR: [Example Branch](https://github.com/stan-dev/math/compare/feature/var-template)
+- Stan Issue: [#1805](https://github.com/stan-dev/math/issues/1805)
+
+# Summary
+[summary]: #summary
+
+This proposes a `static_matrix` and `static_vector` type in the Stan language which on the backend will allow for significant performance opportunities. In Stan Math this will be represented by `var<Eigen::Matrix<double, 1, 1>>` or `var<matrix_cl<double>>`. The implementation will be fully backwards compatible with current Stan code. With optimizations from the compiler the conditions for a `matrix` to be a `static_matrix` can be detected and applied to Stan programs automatically for users.
+
+# Motivation
+[motivation]: #motivation
+
+Currently, a differentiable matrix in Stan is represented as an Eigen matrix holding a pointer to an underlying array of autodiff objects, each of those holding a pointer to the underlying autodiff object implementation. These `N*M` autodiff elements are expensive but necessary so the elements of a matrix can be assigned to without copying the entire matrix. However, in instances where the matrix is treated as a whole object such that underlying subsets of the matrix are never assigned to, we can store one autodiff object representing the entire matrix. 
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+At the language level, `static_matrix` can replace a `matrix` type if subsets of the matrix are never assigned to. That makes code such as the below illegal.
+
+```stan
+static_matrix[N, M] A = // Fill in A...
+A[10, 10] = 5;
```

Comment:
[optional] You can improve clarity by adding a comment that this operation is illegal.

Reply:
The "subsets of the matrix" is confusing.  There are blocks, elements, etc., but "subset" isn't well defined here.  I think what you mean is that we'll be able to assign whole matrices, but won't be able to have any indexing or slicing/blocking on the LHS of an assignment.  I'm not sure how best to say that concisely.

Reply:
"Fill in A" is confusing.  We can have `static_matrix[N, M] A = expression` wherever `expression` is of type `static_matrix` or of type `matrix`.  That is, we should allow the assignment of a regular matrix to a static matrix and vice-versa.

Reply:
>"Fill in A" is confusing

How about "Construct A"?

Reply:
> The "subsets of the matrix" is confusing. There are blocks, elements, etc., but "subset" isn't well defined here. 

I can't think of working here that's better than subset. A subset of a matrix is just a slice, block. element, etc.

> I think what you mean is that we'll be able to assign whole matrices, but won't be able to have any indexing or slicing/blocking on the LHS of an assignment. I'm not sure how best to say that concisely.

I'd actually rather not allow assignment of whole matrices after first declaration specifically because of the naming issues we've had on this design doc. imo I'd rather keep it very simple and then we can be fine calling it `const matrix` or `fixed_matrix`. Saying, "This is constant except when you assign to the entire matrix" also feels a bit odd

Reply:
That's what is great about 'whole matrix' and 'whole vector' as types. The error message is, 'you can only assign the whole matrix/whole vector!'.

I think submatrix is the thing. I guess a subset of a matrix could be element (1, 1), element (7, 5) and any values < 3. There's a definition on Wolfram so good enough!

Reply:
I'm OK with "submatrix" but not with "subset" because these objects are not subsets.

> Saying, "This is constant except when you assign to the entire matrix" also feels a bit odd.

The C distinction between a [constant pointer and a pointer to a constant](https://stackoverflow.com/questions/21476869/constant-pointer-vs-pointer-to-constant) has been confusing programmers for years and I can see how trying to make this kind of subtle distinction about `const` would be confusing.  That's why I don't want to call it "const"!